File: lib/models/transaction_model.dart
class Transaction {
  final String id;
  final String userId;
  final String type;
  final double amount;
  final String? description;
  final String status;
  final String? receiverAccount;
  final String? billType;
  final String? provider;
  final String? customerNumber;
  final DateTime createdAt;
  final DateTime updatedAt;

  Transaction({
    required this.id,
    required this.userId,
    required this.type,
    required this.amount,
    this.description,
    required this.status,
    this.receiverAccount,
    this.billType,
    this.provider,
    this.customerNumber,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Transaction.fromJson(Map<String, dynamic> json) {
    return Transaction(
      id: json['_id'] ?? '',
      userId: json['userId'] ?? '',
      type: json['type'] ?? '',
      amount: (json['amount'] ?? 0.0).toDouble(),
      description: json['description'],
      status: json['status'] ?? 'pending',
      receiverAccount: json['receiverAccount'],
      billType: json['billType'],
      provider: json['provider'],
      customerNumber: json['customerNumber'],
      createdAt: DateTime.parse(json['createdAt'] ?? DateTime.now().toIso8601String()),
      updatedAt: DateTime.parse(json['updatedAt'] ?? DateTime.now().toIso8601String()),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'userId': userId,
      'type': type,
      'amount': amount,
      'description': description,
      'status': status,
      'receiverAccount': receiverAccount,
      'billType': billType,
      'provider': provider,
      'customerNumber': customerNumber,
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
    };
  }
}

class BillDetails {
  final String provider;
  final String customerNumber;
  final String? billType;

  BillDetails({
    required this.provider,
    required this.customerNumber,
    this.billType,
  });

  factory BillDetails.fromJson(Map<String, dynamic> json) {
    return BillDetails(
      provider: json['provider'],
      customerNumber: json['customerNumber'],
      billType: json['billType'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'provider': provider,
      'customerNumber': customerNumber,
      'billType': billType,
    };
  }
}

class ChequeDetails {
  final String chequeNumber;
  final String bankBranch;

  ChequeDetails({
    required this.chequeNumber,
    required this.bankBranch,
  });

  factory ChequeDetails.fromJson(Map<String, dynamic> json) {
    return ChequeDetails(
      chequeNumber: json['chequeNumber'],
      bankBranch: json['bankBranch'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'chequeNumber': chequeNumber,
      'bankBranch': bankBranch,
    };
  }
} 

File: src/models/transaction.model.js
const mongoose = require('mongoose');

const transactionSchema = new mongoose.Schema({
  transactionType: {
    type: String,
    enum: ['transfer', 'bill_payment', 'mobile_recharge', 'cheque_withdrawal', 'deposit'],
    required: true
  },
  amount: {
    type: Number,
    required: true
  },
  sender: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  receiver: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User'
  },
  status: {
    type: String,
    enum: ['pending', 'completed', 'failed'],
    default: 'pending'
  },
  description: {
    type: String
  },
  // For bill payments and mobile recharge
  billDetails: {
    provider: String,
    customerNumber: String,
    billType: {
      type: String,
      enum: ['electricity', 'gas', 'internet', 'mobile']
    }
  },
  // For cheque withdrawals
  chequeDetails: {
    chequeNumber: String,
    bankBranch: String
  },
  reference: {
    type: String,
    unique: true
  }
}, {
  timestamps: true
});

// Generate unique reference number
transactionSchema.pre('save', function(next) {
  if (this.isNew) {
    // Generate reference number: TR + timestamp + random 4 digits
    this.reference = 'TR' + Date.now() + Math.floor(Math.random() * 10000).toString().padStart(4, '0');
  }
  next();
});

const Transaction = mongoose.model('Transaction', transactionSchema);

module.exports = Transaction; 